keel-slack/src/main/kotlin/com/netflix/spinnaker/keel/slack/handlers/GitDataGenerator.kt

```diff
@@ -0,0 +1,65 @@
+package com.netflix.spinnaker.keel.slack.handlers
+
+import com.netflix.spinnaker.keel.api.ScmInfo
+import com.netflix.spinnaker.keel.api.artifacts.GitMetadata
+import com.netflix.spinnaker.keel.api.artifacts.PublishedArtifact
+import com.netflix.spinnaker.keel.artifacts.getScmBaseLink
+import com.slack.api.model.kotlin_extension.block.SectionBlockBuilder
+import org.apache.logging.log4j.util.Strings
+import org.springframework.beans.factory.annotation.Value
+import org.springframework.stereotype.Component
+
+@Component
+class GitDataGenerator(
+  private val scmInfo: ScmInfo,
+  @Value("\${spinnaker.baseUrl}") private val spinnakerBaseUrl: String
+) {
+
+  fun generateRepoLink(gitMetadata: GitMetadata?): String {
+    if (gitMetadata != null) {
+      val baseScmUrl = gitMetadata.commitInfo?.link?.let { getScmBaseLink(scmInfo, it) }
+      return "$baseScmUrl/projects/${gitMetadata.project}/repos/${gitMetadata.repo?.name}"
```

Comment:
This doesn't make a lot of sense to me because, even though you are determining the base URL dynamically, the rest of the URL is hard-coded (presumably for Stash?). This kind of function should be defined in an interface/abstract class and implemented for each SCM type.

Reply:
will renaming it to stash link would help? I don't think of a way (or reason, for now) to make it more generic.

Reply:
It's just that it's inconsistent with the fact that we keep everything else about SCM providers generic. This wouldn't work for anyone trying to use this code from the OSS community unless they happen to be using Stash, for instance. Having said that, given that we're by far the primary users of keel (if not the only ones), I'd say the priority of refactoring is lower and leave it up to you to decide when to address it.

Reply:
P.S. At the very least, I'd recommend adding a `TODO` stating that this only works for Stash at the moment.